### ISSUE TITLE
Recognize side effects when using RxJS

### ISSUE THREAD

**Issue by daflodedeing** [SonarSource/SonarJS#3530]
Hey there, 

is it possible to recognize the possibly bad side effects in the code snippet below:

```js
import './style.css';

import { of, take, switchMap, from, tap } from 'rxjs';

let myTimeout = null;


of('start').pipe(
  take(1),
  switchMap(input =>
    {
      return from(['inner stuff']).pipe(
        tap(_ => { 
          myTimeout = setTimeout(_ => console.log('first tap: timed out'), 10000)
        })
      )
    }
  ),
  tap(_ => console.log('second tap: tap was reached before timeout'))
).subscribe()

```
Imagine the inner function of the first tap operator manipulates something that is then used in the second tap operator. For example a divisor whose initial value is 0 or a database connection to which the inner function of the second tap operator writes to.

As the inner function of the first tap operator does not necessarily run before the inner function of the second tap operator a crash can happen.

Is it possible to get a warning from SonarSource about side effects like that?

**Comment by vilchik-elena** [comment]
Hey @daflodedeing 

Thank you for this idea. Unfortunately as of now there is no rule to detect that, but we might consider this feature next year.

**Comment by daflodedeing** [comment]
Hey @vilchik-elena 

is it technically possible? Do you have similar rules (maybe not rxjs related) to recognize side effects?

**Comment by Wohops** [comment]
Migrated to Jira. New ticket number: JS-111
